### PR TITLE
Add warnings & disable PKCS #7 by default

### LINKED ISSUE #6692: PKCS7 API does not verify certificates
### Summary

The PKCS7 API does not verify the certificates: anyone can include a self-signed certificate and have it be accepted as valid.  Unless an application does additional checks on the certificate, a valid PKCS7 signature provides no security whatsoever.

### System information

Mbed TLS version (number or commit id): 72bffe02b7953e952b8aa8afcf1cbb9ecf5dcf1d
Operating system and version: All
Configuration (if not default, please attach `mbedtls_config.h`): All including PKCS7
Compiler and options (if you used a pre-built binary, please indicate how you obtained it): All
Additional environment information: N/A

### Expected behavior

The PKCS7 API verifies that the certificate has been signed by a trusted authority and not revoked.  The certificate is also checked to be usable for the intended purpose, perhaps by being required to have an Extended Key Usage provided by the caller.

### Actual behavior

The PKCS7 API does not verify the certificate.

### Steps to reproduce

See above.

### Additional information

In my opinion, this makes the current API almost unusable, as it is insecure-by-default.

### PR DESCRIPTION
**Pull request by daverodgman**
Disable PKCS 7 by default; add warnings to `mbedtls_config.h` and `pkcs7.h`.

Also clarify documentation for `mbedtls_pkcs7_signed_xxx_verify()` - this fixes #6692 .

## Gatekeeper checklist

- [x] **changelog** provided, or not required
- [x] **backport** done, or not required
- [x] **tests** provided, or not required

### CODE REVIEW

**Comment by tom-cosgrove-arm** [comment]
Are we going to remove the ChangeLog entry?

File: include/mbedtls/pkcs7.h [hunk 1]
```diff
@@ -220,7 +247,7 @@ int mbedtls_pkcs7_signed_data_verify( mbedtls_pkcs7 *pkcs7,
  * \note           This function is different from mbedtls_pkcs7_signed_data_verify()
  *                 in a way that it directly recieves the hash of the data.
  *
- * \return         A negative error code on failure.
+ * \return         0 if the signatures match, or a negative error code on failure.
```
Comment (tom-cosgrove-arm) [inline include/mbedtls/pkcs7.h:250]: There's really only one signature
```suggestion
 * \return         0 if the signature verifies, or a negative error code on failure.
```

File: include/mbedtls/pkcs7.h [hunk 2]
```diff
@@ -202,15 +218,26 @@ int mbedtls_pkcs7_parse_der( mbedtls_pkcs7 *pkcs7, const unsigned char *buf,
  * \note           This function internally calculates the hash on the supplied
  *                 plain data for signature verification.
  *
- * \return         A negative error code on failure.
+ * \return         0 if the signatures match, or a negative error code on failure.
```
Comment (tom-cosgrove-arm) [inline include/mbedtls/pkcs7.h:221]: We normally talk about the signature (singular) verifying
```suggestion
 * \return         0 if the signature verifies, or a negative error code on failure.
```

**Review by gilles-peskine-arm — APPROVED** [review]
I wouldn't mind making the warning stronger, especially in `mbedtls_config.h` which I fear suggests only a verify-doesn't-enforce-all-policies kind of not ready, whereas we have a parser-risks-memory-corruption level of not ready. But I'm ok with the current wording.

**Review by tom-cosgrove-arm — APPROVED** [review]
I would also prefer a stronger warning, but not going to block this

File: include/mbedtls/mbedtls_config.h [hunk 3]
```diff
@@ -2817,6 +2817,9 @@
 /**
  * \def MBEDTLS_PKCS7_C
  *
+ * This feature is a work in progress and not ready for production. The API may
+ * change. Testing and validation is incomplete.
```
Comment (tom-cosgrove-arm) [inline include/mbedtls/mbedtls_config.h:2821]: Possible suggestion
```suggestion
 * This feature is a work in progress and not ready for production. Testing and
 * validation is incomplete, and handling of malformed inputs may not be robust.
 * The API may change.
```

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Review by gilles-peskine-arm — APPROVED** [review]

**Comment by daverodgman** [comment]
The only CI failures are ABI compatibility, because PKCS 7 is now disabled by default.

**Pull request merged**